### DISCUSSION TITLE
Klipper support

### DISCUSSION THREAD

**Discussion by feelfreelinux** [feelfreelinux/octo4a#59] [General]
Starting with 1.0.4 release, octo4a supports Klipper via the [extension's system](https://github.com/feelfreelinux/octo4a/wiki/Extensions-system). 

**Please note that the Klipper support is currently in it's early testing stages**

## Flashing Klipper

Due to the way Android system works, it is not possible for octo4a to handle the flashing part of Klipper setup. Either use octo4a with your already Klipper-flashed board, or follow the [Klipper's official building instructions](https://www.klipper3d.org/Installation.html#building-and-flashing-the-micro-controller) on your PC / other machine. 

## Setting up Klipper in octo4a

First, please make sure that you have the extra dependencies installed by clicking "Install plugin extras" in octo4a's settings. Later, either connect through SSH or enable the "Remote terminal" option in extension settings.

![image](https://user-images.githubusercontent.com/11230853/140610646-4c5458a1-0a42-45ad-b9a9-698423709824.png)

After that, you can access the web terminal under `phone_ip:5002`. Login: `root`. password: `octoprint` ( or ssh password if you have it set up).

After that run:
`bash -c "$(curl -fsSL https://raw.githubusercontent.com/feelfreelinux/octo4a/master/scripts/setup-klipper.sh)"`. This command will guide you through installation of Klipper, and a Klipper option should appear in the "extensions" tab of octo4a.

## Installing OctoKlipper, configuration

There's an additional step that you need to take to use OctoPrint with Klipper. Go to the plugins options in OctoPrint, and install "OctoKlipper" addon. After installation, restart the server. After restarting, a Klipper tab should appear and everything should be almost good to go.

### Required config changes

In order to use your klipper config with octo4a, you need to change the `[mcu]` serial part to `/dev/ttyOcto4a`. After changing that, and restarting the app everything should connect successfuly.

**Comment by keymaster777** [comment]
I got Klipper working yesterday and wanted to say thanks, this really cleared up the installation process and it works great. This is a great app and contribution to the community.

**Comment by thepowersource** [comment]
I have installed Octo4A and OctoKlipper on an LG G6 and I flashed the Klipper firmware to my SW-X1 using KIAUH from a virtual Debian install on my Mac. I followed the guide here and changed the serial port to /dev/ttyOcto4a but I get this error when trying to connect:

Changing monitoring state from "Offline" to "Opening serial connection"
Connecting to port /dev/ttyOcto4a, baudrate 250000
Changing monitoring state from "Opening serial connection" to "Connecting"
Connected to: Serial<id=0x75e3163910, open=True>(port='/dev/ttyOcto4a', baudrate=250000, bytesize=8, parity='N', stopbits=1, timeout=10.0, xonxoff=False, rtscts=False, dsrdtr=False), starting monitor
Send: N0 M110 N0*125
No answer from the printer within the connection timeout, trying another hello
Send: N0 M110 N0*125
There was a timeout while trying to connect to the printer
Changing monitoring state from "Connecting" to "Offline"
Connection closed, closing down monitor

What I am doing wrong?

**Reply by Dustmuffins** [reply]
Hi. I'm having the same issue. Were you ever able to get this to work?

Same issue on my Galaxy S9 and Sovol SV06.

**Comment by futhgar** [comment]
Has anyone gotten this to work? I installed Klipper FW for a Ender3 V2 foudn here: https://github.com/Klipper3d/klipper/blob/master/config/printer-creality-ender3-v2-2020.cfg

I proceeded to connect the printer and edit the config file within Octoprint in the editor to add /dev/ttyOcto4a. I saved and restarted klipper on Octoprint. I get no response. I'm connected via USB from the phone to the printer.

**Reply by Alessi-io** [reply]
I'm using the same config file as you and it does not work

**Reply by futhgar** [reply]
I wasn't ever able to get it running so I decided to convert an old laptop to a ubuntu server and now I have everything running there, klipper, mainsail, webcam. 

Would've been nice to have gotten this to work by using an old phone instead T_T.

**Reply by Kelsch** [reply]
That sucks that you weren't able to get it working. There are a couple of videos that I made, sorry if they aren't taht high of quality, showing how to get this working. 
https://www.youtube.com/@DalTronPrinting/videos

**Reply by thigo94** [reply]
i've had a similar issue with my ender 3, but after octo4a 1.2.0 i was finally able to make it work (in two devices i previously couldn't). following the guide from DalTron Printing on youtube after i installed klipper on the device via ssh i went to android/data/com.octo4a/files (this names may vary) and substituted the printer cfg directly on the android device. i connected the printer via otg, started the service in octo 4a, went into octoprint on the pc, clicked on connect and got a "?" under firmware, after that i just went to the klipper tab and clicked firmware reset. the printer was identifyed and is working perfectly. if you have any issues try messing with the usb settings on octoprint. if i can do anything else to help you let me know.

**Comment by 4jesse** [comment]
I'm installing this to my Android laptop through ssh and use the command (updated url):
> bash -c "$(curl -jsSL https://github.com/feelfreelinux/octo4a/tree/master/scripts/setup-klipper.sh)"

and I get back this error: 
> bash: /bin/bash: Argument list too long

**Comment by ScriptUp** [comment]
this is closest i got to get klipper runing on phone. ttyOcto4a does not exist in my system. how can i find corect line ? also what ftp client and protocol to use to get inside ssh file system?

**Comment by Zbig667** [comment]
can something be done in regards to:
{
  "error": "Command for custom:octoklipper_restart failed: 500 Internal Server Error: Command for custom:octoklipper_restart failed with return code 1:\n\nSTDOUT:\n\n\nSTDERR:\n * service: service `klipper' does not exist\n"
}

unable to reboot klipper form octoprint...

**Comment by pablunsky** [comment]
If someone is trying this in 2024, I would like to share some difficulties I faced when configuring my printer:

* Everything worked in the tutorial until I had to connect to the printer, I was getting a M112 error without much info in the logs. Looking into Klipper docs I found that you need to update settings in Octoprint to connect to Klipper after installing Octoklipper: https://www.klipper3d.org/Installation.html#configuring-octoprint-to-use-klipper
* After applying these settings, I started getting "Printer not ready" errors, so I had to reflash Klipper in my printer and specify the baud rate for my board (Klipper defaults to 250000 and I needed 115200), and also specify it in printer.cfg

```
[mcu]
serial: ttyOcto4a
baud: 115200
```

I hope this helps someone, and thanks to the team of Octo4a for all their work. My printer is a stock CR10v2.

**Comment by esquivel4x4** [comment]
Hi
Thanks
I have a problem connecting klipper with printer Artillery X2, I made the change in mcu by /dev/ttyOcto4a, but it gives me an error in mcu and does not connect , Error reported by printer
Your printer's firmware reported an error. Due to that OctoPrint will disconnect. Reported error: mcu 'mcu': Unable to connect

**Comment by eredwan** [comment]
Mine is connected. Says "Printer is Ready" when requesting get status. Now im curious, how would I connect a BigTreetech ADXL345 to get this really calibrated. I am using a Amazon Fire Tablet 9th generation. Anyone have any ideas?

**Reply by leonardola** [reply]
Did you get it to work?

**Comment by DougJoseph** [comment]
Hello. I'm trying to install Klipper on some Amazon Fire7 tablets for Octo4A. When I get to this step:

bash -c "$(curl -fsSL https://raw.githubusercontent.com/feelfreelinux/octo4a/master/scripts/setup-klipper.sh)"

The process halts part way through saying:

> Found existing installation: packaging 20.9
> ERROR: Cannot uninstall 'packaging'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.

I did some searching and in flailing about, I tried adding " --ignore-installed packaging" as shown here:

bash -c "$(curl -fsSL https://raw.githubusercontent.com/feelfreelinux/octo4a/master/scripts/setup-klipper.sh)" --ignore-installed packaging

... But that did not work.

Does anyone know how I can get around this?

**Reply by DougJoseph** [reply]
THE SOLUTION I found... 

It dawned on me that my install of Octoprint for Android (Octo4a) was not the newest. I was on Octo4a build 1.0.4.
So, I uninstalled, and installed fresh, which got me to Octo4a build 1.2.5.
Then I re-did the tutorial steps above, and the script process did not halt.   

I now have "Klipper installed!"

**Comment by DougJoseph** [comment]
I followed the steps of the tutorial as closely as I could and now Octoprint cannot start, with a log message like this:

`octoprint could not initialize event manager: invalid version /usr/lib/python3.9/site`

Does anyone have any tips for what to do now?

**Reply by DougJoseph** [reply]
Here is the SOLUTION I found:

It dawned on me that my install of Octoprint for Android (Octo4a) was not the newest. I was on Octo4a build 1.0.4.
So, I uninstalled, and installed fresh, which got me to Octo4a build 1.2.5.

Then I re-ran the steps of this tutorial, and now Octoprint runs.

**Comment by DougJoseph** [comment]
Hmm. I have another Android tablet (Amazon Fire7, that has been rooted) just like the first one I successfully installed Octo4a and Klipper onto, but this one hits an error during the Klipper install script:

```
localhost:~# bash -c "$(curl -fsSL https://raw.githubusercontent.com/feelfreelinux/octo4a/master/scripts/setup-klipper.sh)"
Setting up klipper
Do you have "Plugin extras" installed? (y/n): y
Installing dependencies...

OK: 402 MiB in 153 packages
Collecting python-can
  Using cached python_can-4.3.1-py3-none-any.whl (262 kB)
Requirement already satisfied: wrapt~=1.10 in /usr/lib/python3.10/site-packages (from python-can) (1.15.0)
Collecting msgpack~=1.0.0
  Using cached msgpack-1.0.8.tar.gz (167 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: packaging>=23.1 in /usr/lib/python3.10/site-packages (from python-can) (23.2)
Requirement already satisfied: typing-extensions>=3.10.0.0 in /usr/lib/python3.10/site-packages (from python-can) (4.10.0)
Building wheels for collected packages: msgpack
  Building wheel for msgpack (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for msgpack (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-armv7l-cpython-310
      creating build/lib.linux-armv7l-cpython-310/msgpack
      copying msgpack/__init__.py -> build/lib.linux-armv7l-cpython-310/msgpack
      copying msgpack/exceptions.py -> build/lib.linux-armv7l-cpython-310/msgpack
      copying msgpack/ext.py -> build/lib.linux-armv7l-cpython-310/msgpack
      copying msgpack/fallback.py -> build/lib.linux-armv7l-cpython-310/msgpack
      running build_ext
      building 'msgpack._cmsgpack' extension
      creating build/temp.linux-armv7l-cpython-310
      creating build/temp.linux-armv7l-cpython-310/msgpack
      gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Os -fomit-frame-pointer -g -O2 -Os -fomit-frame-pointer -g -O2 -Os -fomit-frame-pointer -g -O2 -DTHREAD_STACK_SIZE=0x100000 -fPIC -I. -I/usr/include/python3.10 -c msgpack/_cmsgpack.cpp -o build/temp.linux-armv7l-cpython-310/msgpack/_cmsgpack.o
      gcc: fatal error: cannot execute 'cc1plus': execvp: No such file or directory
      compilation terminated.
      error: command 'gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for msgpack
Failed to build msgpack
ERROR: Could not build wheels for msgpack, which is required to install pyproject.toml-based projects
```  

Anyone have any advice for this issue?

**Reply by nicstoun** [reply]
Yes i find way ..after write this comands
python3 -m venv .venv
source .venv/bin/activate
python3 -m pip install -r requirements.txt

then you  can start klipper installation

**Comment by jstillwell** [comment]
Thanks for making this. I recently got my Longer LK5 Pro working with octo4a and Klipper with no small amount of help from this post and some of the responses. The missing piece for me was the `restart_method: command` in the `[mcu]` section. 

```
[mcu]
serial: /dev/ttyOcto4a
restart_method: command
baud: 115200
```

Before I added that line it would appear to connect but show a ? and then would return an error about the mcu section. I looked at the logs and saw something like "restart command failed: not found" and that, plus google, led me to `restart_method: command`.

Maybe this will help someone else in a similar situation.

**Comment by VeridisQuo555** [comment]
in having the issue that when I type this script bash -c "$(curl -fsSL https://raw.githubusercontent.com/feelfreelinux/octo4a/master/scripts/setup-klipper.sh)" to download klipper i get stuck on 1/7

**Comment by jstillwell** [comment]
No, I moved to klipper and other tools.

-John

**Comment by aracchi9** [comment]
Im having the following error when executing the bash command:

```
error: externally-managed-environment

× This environment is externally managed
╰─> 
    The system-wide python installation should be maintained using the system
    package manager (apk) only.
    
    If the package in question is not packaged already (and hence installable via
    "apk add py3-somepackage"), please consider installing it inside a virtual
    environment, e.g.:
    
    python3 -m venv /path/to/venv
    . /path/to/venv/bin/activate
    pip install mypackage
    
    To exit the virtual environment, run:
    
    deactivate
    
    The virtual environment is not deleted, and can be re-entered by re-sourcing
    the activate file.
    
    To automatically manage virtual environments, consider using pipx (from the
    pipx package).

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

Does anyone know what could be happening?

**Reply by aracchi9** [reply]
I was able to fix this but now im not seeing klipper in the Octo4A extensions menu
